lib/kafka/protocol/message.rb

```diff
@@ -57,11 +57,12 @@ def decompress
         data = codec.decompress(value)
         message_set_decoder = Decoder.from_string(data)
         message_set = MessageSet.decode(message_set_decoder)
+        base_offset = offset - message_set.size + 1
 
         # The contained messages need to have their offset corrected.
         messages = message_set.messages.each_with_index.map do |message, i|
```

Comment:
`each_with_index` and `+i` will not work for compacted topics. should use the real relative offsets instead.

Ref: 
https://cwiki.apache.org/confluence/display/KAFK/KIP-31+-+Move+to+relative+offsets+in+compressed+message+sets

> When the log cleaner compacts log segments, if multiple message sets are compacted into one message set, the broker needs to update the inner message's relative offset values. (This will leave "holes" inside the new wrapped message).